Nil pointer error when a SecurityPolicy refers to a UDS Backend

```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: Backend
metadata:
  name: ext-auth-svc
spec:
  endpoints:
  - unix:
      path: /var/run/envoy-uds/ext-auth.sock
---

apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: ext-auth-poc
spec:
  extAuth:
    failOpen: false
    http:
      backendRefs:
      - group: gateway.envoyproxy.io
        kind: Backend
        name: ext-auth-svc
      headersToBackend:
      - x-current-user
    recomputeRoute: true
  targetRefs:
  - group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: ext-auth-poc
```

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2768705]

goroutine 305 [running]:
github.com/envoyproxy/gateway/internal/gatewayapi.backendRefAuthority(0xc0011169c0?, 0xc002a781e8, 0x7?)
	/home/runner/work/gateway/gateway/internal/gatewayapi/securitypolicy.go:960 +0x185
github.com/envoyproxy/gateway/internal/gatewayapi.(*Translator).buildExtAuth(0x78f5d88?, 0xc0010c9380, 0xc002e28c88, 0xc003199140?)
	/home/runner/work/gateway/gateway/internal/gatewayapi/securitypolicy.go:908 +0x3b6
github.com/envoyproxy/gateway/internal/gatewayapi.(*Translator).translateSecurityPolicyForRoute(0xc002a798f8, 0xc0010c9380, {0x78f5d88, 0xc002ca6bc0}, 0xc002e28c88, 0xc000ad7290)
	/home/runner/work/gateway/gateway/internal/gatewayapi/securitypolicy.go:373 +0x677
github.com/envoyproxy/gateway/internal/gatewayapi.(*Translator).ProcessSecurityPolicies(0xc002a798f8, {0xc0000cd598, 0x1, 0x1}, {0xc00333e428, 0x1, 0x1}, {0xc003199570, 0x1, 0x1}, ...)
	/home/runner/work/gateway/gateway/internal/gatewayapi/securitypolicy.go:150 +0x1931
github.com/envoyproxy/gateway/internal/gatewayapi.(*Translator).Translate(0xc002a798f8, 0xc002e28c88)
	/home/runner/work/gateway/gateway/internal/gatewayapi/translator.go:217 +0xaa5
github.com/envoyproxy/gateway/internal/gatewayapi/runner.(*Runner).subscribeAndTranslate.func1({{0xc000135e30?, 0x0?}, 0x2?, 0xc0031a65b8?}, 0xc001039490)
	/home/runner/work/gateway/gateway/internal/gatewayapi/runner/runner.go:167 +0x42c
github.com/envoyproxy/gateway/internal/message.HandleSubscription[...]({{0x3254be7, 0xb?}, {0x3266e8a?, 0x12?}}, 0xc0010351f0?, 0xc002a79f98)
	/home/runner/work/gateway/gateway/internal/message/watchutil.go:76 +0x1118
github.com/envoyproxy/gateway/internal/gatewayapi/runner.(*Runner).subscribeAndTranslate(0xc002315200, {0x78babb0?, 0xc001202000?})
	/home/runner/work/gateway/gateway/internal/gatewayapi/runner/runner.go:119 +0x7e
created by github.com/envoyproxy/gateway/internal/gatewayapi/runner.(*Runner).Start in goroutine 137
	/home/runner/work/gateway/gateway/internal/gatewayapi/runner/runner.go:85 +0x2a5
```

--- a/internal/gatewayapi/securitypolicy.go
+++ b/internal/gatewayapi/securitypolicy.go
@@ -962,10 +962,16 @@ func backendRefAuthority(resources *resource.Resources, backendRef *gwapiv1.Back
 		}
 	}
 
-	return net.JoinHostPort(
-		fmt.Sprintf("%s.%s", backendRef.Name, backendNamespace),
-		strconv.Itoa(int(*backendRef.Port)),
-	)
+	// Port is mandatory for Kubernetes services
+	if backendKind == resource.KindService {
+		return net.JoinHostPort(
+			fmt.Sprintf("%s.%s", backendRef.Name, backendNamespace),
+			strconv.Itoa(int(*backendRef.Port)),
+		)
+	}
+
+	// Fallback to the backendRef name, normally it's a unix domain socket in this case
+	return fmt.Sprintf("%s.%s", backendRef.Name, backendNamespace)
 }
 
 func (t *Translator) buildAuthorization(policy *egv1a1.SecurityPolicy) (*ir.Authorization, error) {

--- a/internal/gatewayapi/testdata/securitypolicy-with-extauth-backend.in.yaml
+++ b/internal/gatewayapi/testdata/securitypolicy-with-extauth-backend.in.yaml
@@ -107,6 +107,15 @@ backends:
         - fqdn:
             hostname: 'primary.foo.com'
             port: 3000
+  - apiVersion: gateway.envoyproxy.io/v1alpha1
+    kind: Backend
+    metadata:
+      name: backend-uds
+      namespace: default
+    spec:
+      endpoints:
+        - unix:
+            path: '/var/run/uds.sock'
 referenceGrants:
   - apiVersion: gateway.networking.k8s.io/v1alpha2
     kind: ReferenceGrant
@@ -179,7 +188,6 @@ securityPolicies:
       extAuth:
         http:
           backendRef:
-            name: backend-fqdn
+            name: backend-uds
             kind: Backend
             group: gateway.envoyproxy.io
-            port: 3000

--- a/internal/gatewayapi/testdata/securitypolicy-with-extauth-backend.out.yaml
+++ b/internal/gatewayapi/testdata/securitypolicy-with-extauth-backend.out.yaml
@@ -17,6 +17,23 @@ backends:
       reason: Accepted
       status: "True"
       type: Accepted
+- apiVersion: gateway.envoyproxy.io/v1alpha1
+  kind: Backend
+  metadata:
+    creationTimestamp: null
+    name: backend-uds
+    namespace: default
+  spec:
+    endpoints:
+    - unix:
+        path: /var/run/uds.sock
+  status:
+    conditions:
+    - lastTransitionTime: null
+      message: The Backend was accepted
+      reason: Accepted
+      status: "True"
+      type: Accepted
 gateways:
 - apiVersion: gateway.networking.k8s.io/v1
   kind: Gateway
@@ -322,8 +339,7 @@ securityPolicies:
         backendRef:
           group: gateway.envoyproxy.io
           kind: Backend
-          name: backend-fqdn
-          port: 3000
+          name: backend-uds
     targetRef:
       group: gateway.networking.k8s.io
       kind: HTTPRoute
@@ -526,14 +542,15 @@ xdsIR:
         security:
           extAuth:
             http:
-              authority: primary.foo.com:3000
+              authority: backend-uds.default
               destination:
                 name: securitypolicy/default/policy-for-http-route-3-http-backendref/extauth/0
                 settings:
-                - addressType: FQDN
+                - addressType: IP
                   endpoints:
-                  - host: primary.foo.com
-                    port: 3000
+                  - host: ""
+                    path: /var/run/uds.sock
+                    port: 0
                   protocol: HTTP
                   weight: 1
               path: ""

--- a/release-notes/current.yaml
+++ b/release-notes/current.yaml
@@ -20,6 +20,7 @@ new features: |
 
 # Fixes for bugs identified in previous versions.
 bug fixes: |
+  Fixed a nil pointer error that occurs when a SecurityPolicy refers to a UDS backend
 
 # Enhancements that improve performance.
 performance improvements: |

